In [1]:
%matplotlib inline
import os
from pathlib import Path

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
from tqdm import trange, tqdm

from sklearn.ensemble import GradientBoostingClassifier, RandomForestClassifier
from sklearn.linear_model import LinearRegression
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import train_test_split, StratifiedKFold
from sklearn.preprocessing import StandardScaler
from sklearn.utils import shuffle
import lightgbm as lgb

from scipy.signal import hilbert
from scipy.signal import hann
from scipy.signal import convolve
from scipy import stats

from fastai.datasets import Config

import warnings
warnings.filterwarnings("ignore")

base_path = Config.data_path()

# Competiton files setup

In [2]:
data_path = base_path/'LANL_Earthquake_Prediction'
competition_name = 'LANL-Earthquake-Prediction'
data_path

PosixPath('/home/jupyter/.fastai/data/LANL_Earthquake_Prediction')

In [3]:
train_prcessed_segments_path = data_path/'train'/'processed_df'
test_prcessed_segments_path = data_path/'test_processed_df'

In [4]:
train_prcessed_segments_path.ls()

[PosixPath('/home/jupyter/.fastai/data/LANL_Earthquake_Prediction/train/processed_df/no_overlap_only_time_features.csv'),
 PosixPath('/home/jupyter/.fastai/data/LANL_Earthquake_Prediction/train/processed_df/step_10000_only_time_features.csv')]

In [5]:
test_prcessed_segments_path.ls()

[PosixPath('/home/jupyter/.fastai/data/LANL_Earthquake_Prediction/test_processed_df/only_time_features.csv')]

In [89]:
# training_set_step_10000_only_time = pd.read_csv(train_prcessed_segments_path/'step_10000_only_time_features.csv')
training_set_no_overlap_only_time = pd.read_csv(train_prcessed_segments_path/'no_overlap_only_time_features.csv')

In [90]:
test_set_only_time = pd.read_csv(test_prcessed_segments_path/'only_time_features.csv')

In [9]:
test_set_only_time.head()

name      mean        std    max    min  mean_change_abs  \
0  seg_1fe311  3.824900   4.124970   99.0  -79.0        -0.000040   
1  seg_9f22a0  4.442760   4.209801  150.0  -84.0         0.000033   
2  seg_8f1127  4.823960  10.636562  271.0 -199.0        -0.000053   
3  seg_eabb2c  4.298393   4.091173  121.0 -121.0        -0.000027   
4  seg_d24095  4.187107   4.005560  108.0  -96.0         0.000100   

   mean_change_rate  abs_max  abs_min  std_first_50000  ...  \
0      75048.795174     99.0      0.0         3.866490  ...   
1      74949.483408    150.0      0.0         3.298534  ...   
2      75032.648534    271.0      0.0         8.777746  ...   
3      74918.297327    121.0      0.0         5.441324  ...   
4      75052.671652    108.0      0.0         4.511398  ...   

   std_roll_mean_1000  max_roll_mean_1000  min_roll_mean_1000  \
0            0.223865               4.432               3.204   
1            0.238105               5.066               3.739   
2            0.227011               7.177               3.732   
3            0.230050               4.894               3.564   
4            0.222505               4.829               3.553   

   q01_roll_mean_1000  q05_roll_mean_1000  q95_roll_mean_1000  \
0               3.289               3.410               4.181   
1               3.862               4.011               4.784   
2               4.323               4.487               5.177   
3               3.734               3.884               4.659   
4               3.702               3.809               4.540   

   q99_roll_mean_1000  av_change_abs_roll_mean_1000  \
0               4.291                 -8.791946e-07   
1               4.930                 -6.577181e-07   
2               5.288                  1.355705e-06   
3               4.778                  3.221477e-07   
4               4.660                  1.328859e-06   

   av_change_rate_roll_mean_1000  abs_max_roll_mean_1000  
0                   74487.329582                   4.432  
1                   74496.457774                   5.066  
2                   74609.066800                   7.177  
3                   74299.608866                   4.894  
4                   74534.006113                   4.829  

[5 rows x 139 columns]

In [10]:
# training_set_no_overlap_only_time.drop(['Unnamed: 0'], axis=1, inplace=True)
training_set_no_overlap_only_time.head()

target      mean       std    max    min  mean_change_abs  \
0  1.430797  4.884113  5.101106  104.0  -98.0        -0.000080   
1  1.391499  4.725767  6.588824  181.0 -154.0         0.000000   
2  1.353196  4.906393  6.967397  140.0 -106.0        -0.000013   
3  1.313798  4.902240  6.922305  197.0 -199.0         0.000000   
4  1.274400  4.908720  7.301110  145.0 -126.0        -0.000007   

   mean_change_rate  abs_max  abs_min  std_first_50000  ...  \
0      74836.577199    104.0      0.0         6.488552  ...   
1      74891.736232    181.0      0.0         7.305233  ...   
2      75099.224451    140.0      0.0         6.104836  ...   
3      74933.991879    199.0      0.0         6.238109  ...   
4      75010.016046    145.0      0.0         5.323830  ...   

   std_roll_mean_1000  max_roll_mean_1000  min_roll_mean_1000  \
0            0.295715               5.629               3.896   
1            0.231587               5.667               3.412   
2            0.267012               5.957               4.055   
3            0.266700               5.858               3.722   
4            0.228005               6.078               3.918   

   q01_roll_mean_1000  q05_roll_mean_1000  q95_roll_mean_1000  \
0               4.072               4.379               5.338   
1               4.233               4.345               5.066   
2               4.239               4.446               5.344   
3               4.304               4.433               5.317   
4               4.434               4.543               5.306   

   q99_roll_mean_1000  av_change_abs_roll_mean_1000  \
0               5.484                     -0.000002   
1               5.223                     -0.000002   
2               5.486                      0.000001   
3               5.453                     -0.000003   
4               5.491                      0.000002   

   av_change_rate_roll_mean_1000  abs_max_roll_mean_1000  
0                   74222.343443                   5.629  
1                   74364.426478                   5.667  
2                   74594.984584                   5.957  
3                   74333.143623                   5.858  
4                   74428.435969                   6.078  

[5 rows x 139 columns]

In [32]:
training_set_step_10000_only_time.drop(['Unnamed: 0'], axis=1, inplace=True)
training_set_step_10000_only_time.head()

target      mean       std    max    min  mean_change_abs  \
0  1.430797  4.884113  5.101106  104.0  -98.0        -0.000080   
1  1.427600  4.857127  4.324007   52.0  -56.0         0.000007   
2  1.425498  4.837627  5.334216  181.0 -154.0        -0.000027   
3  1.423396  4.811160  5.322245  181.0 -154.0         0.000033   
4  1.420198  4.792853  5.418203  181.0 -154.0         0.000027   

   mean_change_rate  abs_max  abs_min  std_first_50000  ...  \
0      74836.577199    104.0      0.0         6.488552  ...   
1      74890.988734     56.0      0.0         4.506360  ...   
2      74986.731761    181.0      0.0         4.680861  ...   
3      74989.309697    181.0      0.0         4.748028  ...   
4      75145.482449    181.0      0.0         4.725781  ...   

   std_roll_mean_1000  max_roll_mean_1000  min_roll_mean_1000  \
0            0.295715               5.629               3.896   
1            0.290901               5.629               3.896   
2            0.298291               5.667               3.412   
3            0.303731               5.667               3.412   
4            0.302770               5.667               3.412   

   q01_roll_mean_1000  q05_roll_mean_1000  q95_roll_mean_1000  \
0               4.072               4.379               5.338   
1               4.072               4.375               5.320   
2               4.069               4.360               5.320   
3               4.069               4.360               5.320   
4               4.069               4.359               5.320   

   q99_roll_mean_1000  av_change_abs_roll_mean_1000  \
0               5.484                 -1.704698e-06   
1               5.483                 -1.906040e-06   
2               5.483                 -6.107383e-07   
3               5.483                 -3.402685e-06   
4               5.483                 -3.221477e-07   

   av_change_rate_roll_mean_1000  abs_max_roll_mean_1000  
0                   74222.343443                   5.629  
1                   74333.053867                   5.629  
2                   74474.087926                   5.667  
3                   74346.540137                   5.667  
4                   74551.377120                   5.667  

[5 rows x 139 columns]

In [86]:
feature_scores_path = data_path/'feature_evaluation'/'single_fueature_scores.csv'
feature_scores = pd.read_csv(feature_scores_path)
# best_22_features = list(feature_scores.sort_values('score')[:22]['feature_name'])

In [77]:
sorted_features = list(feature_scores.sort_values('score')['feature_name'])
print(f"# of features: {len(sorted_features)}")

# of features: 137


In [78]:
best_70_features = sorted_features[:70]
best_22_features = sorted_features[:22]

In [91]:
# Current training and test sets
df_train = training_set_no_overlap_only_time
df_test = test_set_only_time

df_test_segment_names = df_test['name']
df_test.drop(['name'], axis=1, inplace=True)

In [80]:
df_test = df_test[best_22_features]
best_22_features.append('target')
df_train = df_train[best_22_features]
df_train.head()

q05_roll_std_100  q05_roll_std_1000  q05_roll_std_10  q01_roll_std_100  \
0          2.475639           2.706474         1.636392          2.302634   
1          2.475965           2.674879         1.646545          2.300285   
2          2.538591           2.761534         1.686548          2.374613   
3          2.496442           2.716991         1.649916          2.330539   
4          2.491521           2.719174         1.646545          2.314731   

   q01_roll_std_10  q01_roll_std_1000       mad  MA_400MA_std_mean   q95  \
0         1.264911           2.616094  3.263401           4.155806  11.0   
1         1.286684           2.612482  3.574302           4.608579  12.0   
2         1.316561           2.660178  3.948411           5.122560  13.0   
3         1.269296           2.624962  3.647117           4.698889  12.0   
4         1.269296           2.628699  3.826052           4.910515  12.0   

    q95  ...  ave_roll_std_100  ave_roll_std_1000  MA_700MA_std_mean  \
0  11.0  ...          4.050450           4.288590           4.229416   
1  12.0  ...          4.436359           4.843486           4.735352   
2  13.0  ...          4.917334           5.423013           5.283180   
3  12.0  ...          4.533343           4.939280           4.826369   
4  12.0  ...          4.761149           5.121868           5.023478   

   Hilbert_mean  q05   q01   q99  abs_q99  q95_roll_std_1000    target  
0      7.027028 -2.0  -8.0  18.0     20.0           8.185756  1.430797  
1      7.380383 -2.0 -11.0  21.0     24.0          10.544982  1.391499  
2      8.016930 -3.0 -15.0  26.0     30.0          14.845834  1.353196  
3      7.606850 -2.0 -12.0  22.0     26.0          11.715642  1.313798  
4      7.895403 -2.0 -15.0  26.0     32.0          13.923676  1.274400  

[5 rows x 23 columns]

In [33]:
# Switch only train set
df_train = training_set_step_10000_only_time

In [81]:
print(f"Train length: {len(df_train)}")
print(f"Test length: {len(df_test)}")

Train length: 4178
Test length: 2624


We will use LightGBM as a model to predict target for test dataset. This is a tree-based model, which cannot see frequency of the feature (number of occurrences of particular value in a feature). It turns out that this is important in this case, so frequencies needs to be added as separate feature. 

# Model Training
## 1st Layer of ensamble

In [47]:
param = {
    'learning_rate': 0.04,
    'num_leaves': 3,
    'metric':'mean_absolute_error',
    'boost_from_average':'false',
    'feature_fraction': 0.8, # This is important; we must use ALL features in every iteration to make sure that feature freq will be used
    'max_depth': -1,
    'objective': 'regression',
#     'verbosity': -10
}

In [92]:
%%time

from sklearn.model_selection import KFold
from sklearn.metrics import mean_absolute_error
from sklearn.preprocessing import StandardScaler

nbr_of_folds = 5
nbr_of_rounds = 10000

folds = KFold(
    n_splits=nbr_of_folds,
    shuffle=True,
    random_state=2019
)

scaler = StandardScaler()

out_of_fold_predictions = np.zeros(len(df_train))
test_predictions = np.zeros(len(df_test))

train_data = df_train.drop('target', axis=1)
targets = df_train['target']
    
folds_predictions = np.zeros(len(df_test))
training_results = dict()

## Scaling
# scaler.fit(train_data)
# train_data = pd.DataFrame(scaler.transform(train_data))
# df_test = pd.DataFrame(scaler.transform(df_test))
##

for fold_nbr, (train_idx, valid_idx) in enumerate(folds.split(train_data, targets)):
    fold_train_data = train_data.iloc[train_idx]
    fold_train_target = targets.iloc[train_idx]
        
    fold_valid_data = train_data.iloc[valid_idx]
    fold_valid_target = targets.iloc[valid_idx]

    lgb_fold_train = lgb.Dataset(fold_train_data, label=fold_train_target)
    lgb_fold_valid = lgb.Dataset(fold_valid_data, label=fold_valid_target, reference=lgb_fold_train)

    model = lgb.train(
        param, 
        lgb_fold_train, 
        nbr_of_rounds, 
        valid_sets=[lgb_fold_train, lgb_fold_valid], 
       early_stopping_rounds=500,
        verbose_eval=False, 
        evals_result=training_results,
#        verbose_eval=5000
    )
    
#     print(training_results)
    scores = training_results['valid_1']['l1']
    best_score = min(scores)
    best_round = scores.index(min(scores))
#     print(f"Fold #{fold_nbr}: Best l1: {best_score} ({best_round} iteration)")

    out_of_fold_predictions[valid_idx] = model.predict(fold_valid_data, num_iteration=best_round)
    folds_predictions += model.predict(df_test, num_iteration=best_round)

test_predictions = folds_predictions/nbr_of_folds

print(f"Final score: {mean_absolute_error(targets.values, out_of_fold_predictions)}")

Final score: 2.0640357531770532
CPU times: user 11.3 s, sys: 0 ns, total: 11.3 s
Wall time: 2.84 s


In [64]:
test_predictions

array([8.928522, 7.451592, 2.211401, 8.782383, ..., 3.685757, 2.782239, 3.463163, 7.971189])

Best 22 features

Without scaler: 2.07108839172104 (LB 1.614)
With scaler: 2.0674590671564608 

Best 70 features

Without scaler: 2.0766704829894915 (LB 1.604)
With scaler: 2.0746327066782286

All features

Without scaler:  2.0640357531770532 (LB 1.554?)
With scaler: 2.0634338328764064

In [103]:
np.save(data_path/'oof', out_of_fold_predictions)
np.save(data_path/'ensemble_1st_lvl', test_predictions)

## 2nd Layer of ensamble

In [109]:
from sklearn.linear_model import LogisticRegression

X_train, X_valid, Y_train, Y_valid = train_test_split(
    out_of_fold_predictions, 
    targets.values, 
    test_size=0.15, 
    random_state=2019
)

ensamble_second_layer = LogisticRegression(n_jobs=-1)
ensamble_second_layer.fit(X_train, Y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn', n_jobs=-1,
          penalty='l2', random_state=None, solver='warn', tol=0.0001,
          verbose=0, warm_start=False)

In [118]:
probs = ensamble_second_layer.predict_proba(X_valid)
roc_auc_score(Y_valid, probs[:,1])

0.9040947053429751

In [115]:
import statsmodels.api as sm

logr = sm.Logit(targets, out_of_fold_predictions)
logr = logr.fit(disp=0)
ensemble_preds = logr.predict(out_of_fold_predictions)
ensemble_auc = roc_auc_score(targets, ensemble_preds)  
print('##################')
print('Combined Model with magic Val_AUC=',round(ensemble_auc,5))

##################
Combined Model with magic Val_AUC= 0.91402


In [117]:
lgb_ensemble_train = lgb.Dataset(X_train, label=Y_train)
lgb_ensemble_valid = lgb.Dataset(X_valid, label=Y_valid, reference=lgb_ensemble_train)

model = lgb.train(
    param, 
    lgb_ensemble_train, 
    50000, 
    valid_sets=[lgb_ensemble_train, lgb_ensemble_valid], 
    early_stopping_rounds=500,
#     verbose_eval=False, 
#             evals_result=training_results
    verbose_eval=1000
    )

# ensemble_predicions = model.predict(fold_valid_data)
# folds_predictions += model.predict(df_test, num_iteration=best_round)


Training until validation scores don't improve for 500 rounds.
[1000]	training's auc: 0.893676	valid_1's auc: 0.878561
[2000]	training's auc: 0.912475	valid_1's auc: 0.895864
[3000]	training's auc: 0.921207	valid_1's auc: 0.902603
[4000]	training's auc: 0.926821	valid_1's auc: 0.906273
[5000]	training's auc: 0.931289	valid_1's auc: 0.908226
[6000]	training's auc: 0.935035	valid_1's auc: 0.909387
[7000]	training's auc: 0.938219	valid_1's auc: 0.910122
[8000]	training's auc: 0.94106	valid_1's auc: 0.910847
[9000]	training's auc: 0.943547	valid_1's auc: 0.911536
[10000]	training's auc: 0.945812	valid_1's auc: 0.912033
[11000]	training's auc: 0.947898	valid_1's auc: 0.912439
[12000]	training's auc: 0.949817	valid_1's auc: 0.912743
[13000]	training's auc: 0.951595	valid_1's auc: 0.912847
Early stopping, best iteration is:
[12584]	training's auc: 0.950872	valid_1's auc: 0.91289


# Final prediction and submission to Kaggle

In [65]:
submission_df = pd.DataFrame({'seg_id': df_test_segment_names, 'time_to_failure': test_predictions})

In [66]:
submission_df.head()

seg_id  time_to_failure
0  seg_1fe311         8.928522
1  seg_9f22a0         7.451592
2  seg_8f1127         2.211401
3  seg_eabb2c         8.782383
4  seg_d24095         8.343232

In [67]:
submission_file_name = "only_best_70_time_features_lgbm"

submission_path = data_path/'submissions'
submission_path.mkdir(exist_ok=True)
submission_file = submission_path/f"{submission_file_name}.csv"
submission_df.to_csv(submission_file, index=False)

In [68]:
# Only 2 submission allowed per day!
!kaggle competitions submit {competition_name} -f {submission_file} -m "LGBM, with only 70 best time features"

100%|██████████████████████████████████████| 74.8k/74.8k [00:02<00:00, 30.7kB/s]
Successfully submitted to LANL Earthquake Prediction